Keen.NetStandard.Test/EventCacheTest.cs

```diff
@@ -0,0 +1,110 @@
+﻿using Keen.NetStandard.EventCache;
+using NUnit.Framework;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Newtonsoft.Json.Linq;
+
+
+namespace Keen.NetStandard.Test
+{
+    [TestFixture]
+    public class EventCacheTest : TestBase
+    {
+        static readonly object[] Providers = 
+        {
+            new object[] { new EventCacheMemory() }
+        };
+
+        [Test]
+        [TestCaseSource("Providers")]
+        public void AddEvent_Null_Throws(IEventCache cache)
+        {
+            Assert.ThrowsAsync<KeenException>(() => cache.Add(null));
+        }
+
+        [Test]
+        [TestCaseSource("Providers")]
+        public async Task AddEvent_ValidObject_Success(IEventCache cache)
+        {
+            await cache.Add(new CachedEvent("url", JObject.FromObject( new { Property = "Value" })));
+        }
+
+        [Test]
+        [TestCaseSource("Providers")]
+        public async Task AddEvent_AddNotEmpty_Success(IEventCache cache)
+        {
+            await cache.Clear();
+            Assert.Null(await cache.TryTake());
+            await cache.Add(new CachedEvent("url", JObject.FromObject( new { Property = "Value" })));
+            Assert.NotNull(await cache.TryTake());
+        }
+
+        [Test]
+        [TestCaseSource("Providers")]
+        public async Task AddEvent_AddClearEmpty_Success(IEventCache cache)
+        {
+            await cache.Add( new CachedEvent("url", JObject.FromObject( new { Property = "Value" })));
+            await cache.Clear();
+            Assert.Null(await cache.TryTake());
+        }
+
+        [Test]
+        [TestCaseSource("Providers")]
+        public async Task AddEvent_Iterate_Success(IEventCache cache)
+        {
+            await cache.Clear();
+            await cache.Add( new CachedEvent("url", JObject.FromObject( new { Property = "Value" })));
+            await cache.Add( new CachedEvent("url", JObject.FromObject( new { Property = "Value" })));
+            Assert.NotNull(await cache.TryTake());
```

Comment:
Similar to my comment above, in my notes from a while back I have jotted down that I don't like the `TryTake()` nomenclature:
```
Generally the Try*() paradigm in C# would suggest this should be doing something that can throw, but does it without throwing, and instead returns a bool and takes an out/ref param, but you can't do that with async methods, but you can return a Task<Tuple<Thing, bool>>.
```
Though with C#7 that changes a bit since you can have inlined out params and easier syntax for returning multiple values as tuples or unpacked.  Again, not due to this change, but something to keep in mind if there's a potential redesign or opportunity for doing so.